lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -577,15 +577,15 @@ def set_zsort(self, zsort):
 
     def get_vector(self, segments3d):
         """Optimize points for projection."""
-        if len(segments3d):
-            xs, ys, zs = np.row_stack(segments3d).T
-        else:  # row_stack can't stack zero arrays.
-            xs, ys, zs = [], [], []
-        ones = np.ones(len(xs))
-        self._vec = np.array([xs, ys, zs, ones])
-
-        indices = [0, *np.cumsum([len(segment) for segment in segments3d])]
-        self._segslices = [*map(slice, indices[:-1], indices[1:])]
+        if isinstance(segments3d, (np.ndarray, np.ma.MaskedArray)):
+            self._segments = segments3d
+        else:
+            num_faces = len(segments3d)
+            max_verts = max((len(face) for face in segments3d), default=0)
+            padded = np.full((num_faces, max_verts, 3), np.nan)
+            for i, face in enumerate(segments3d):
+                padded[i, :len(face)] = face
+            self._segments = np.ma.masked_invalid(padded)
```

Comment:
Faster as:
```suggestion
            num_faces = len(segments3d)
            max_verts = max((len(face) for face in segments3d), default=0)
            padded = np.empty((num_faces, max_verts, 3))
            mask = np.zeros((num_faces, max_verts, 3), dtype=bool)
            for i, face in enumerate(segments3d):
                padded[i, :len(face)] = face
                mask[i, len(face):] = True
            self._segments = np.ma.array(padded, mask=mask)
```

I'm not convinced masked arrays are the right approach here. Component-wise masking feels a little awkward. Would it be better to store a `_segment_lengths` array separately?

Reply:
It would be a better model of the data, but that wouldn't allow doing vectorized Z-reductions, because there are no NumPy kernels for arrays with lengths.

Reply:
That's a good point.
Perhaps another good option would be to store the mask separately, meaning it could be `(NF, NV)` rather than `(NF, NV, 3)`. If you have numpy >= 1.17 (@anntzer, is that the minimum yet?), then you can use the `where` argument to the reductions directly.

Reply:
Also, the suggestion seems to be at least 1,5x slower in my benchmark (probably due to 2x more Python instructions being executed in total):
```python
Python 3.7.4 (default, Aug 13 2019, 15:17:50)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: def f():
   ...:     x = np.full((10000, 12, 3), np.nan)
   ...:     for i in range(10000):
   ...:         valid = i % 10 + 1
   ...:         x[i, :valid] = 1
   ...:     return np.ma.masked_invalid(x)
   ...:

In [3]: %timeit f()
11.1 ms ± 245 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: def g():
   ...:     x = np.empty((10000, 12, 3))
   ...:     mask = np.zeros((10000, 12, 3), dtype=np.bool)
   ...:     for i in range(10000):
   ...:         valid = i % 10 + 1
   ...:         x[i, :valid] = 1
   ...:         mask[i, valid:] = True
   ...:     return np.ma.MaskedArray(x, mask=mask)
   ...:

In [5]: %timeit g()
19 ms ± 434 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [6]: %timeit g()
18.7 ms ± 148 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: %timeit f()
11.2 ms ± 575 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
Huh, that surprises me a little.

What worries me is that the approach you use here has a behavior change if a `nan` is passed in by the user.

Reply:
Also, use `bool` or `np.bool_`, never `np.bool` which is a soon-to-be-deprecated version of the former.

Reply:
@eric-wieser No, current policy is to bump to >=1.16 in July (1.15+2y) and >=1.17 next January (1.16+2y).

Reply:
> What worries me is that the approach you use here has a behavior change if a nan is passed in by the user. So even if my suggestion is not in fact faster, it's probably still safer.

This comment still applies.

Reply:
NaN vertices are supposed to get ignored while drawing, so that should be fine. On the other hand I think that might mess up the invariant that the last dimension of `mask` yields constant values. Any opinions? Want to relax the invariant or use the slower implementation?

Reply:
What about this: we use my implementation, but in the end we fix-up the invariant by masking out the whole vertex if even a single component is masked?

Reply:
This profiles quite favorably on my machine:
```
def i():
    x = np.empty((10000, 12, 3))

    face_index = np.arange(x.shape[1])

    for i in range(10000):
        valid = i % 10 + 1
        x[i, :valid] = 1
    mask = face_index >= np.array([i % 10 + 1 for i in range(10000)])[:,None]
    mask = np.broadcast_to(mask[...,None], x.shape)
    return np.ma.MaskedArray(x, mask=mask, copy=False)
```

Reply:
So, that would be:
```suggestion
            num_faces = len(segments3d)
            face_lens = np.array([len(face) for face in segments3d], dtype=np.intp)
            # max(initial=0) isn't legal until a newer numpy version
            max_verts = face_lens.max() if len(face_lens) else 0
            padded = np.empty((num_faces, max_verts, 3))
            for i, face in enumerate(segments3d):
                padded[i, :len(face)] = face
            mask = np.arange(max_verts) >= face_lens[:,None]
            mask = mask[..., None]
            # ma.array does not broadcast the mask for us
            mask = np.broadcast_to(mask, padded.shape)
            self._segments = np.ma.array(padded, mask=mask)
```

Reply:
That looks sweet! It runs pretty much as fast as my original version!